src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -902,36 +895,37 @@ fn build_diagnostic(
                         found = link_range.len(),
                     ));
                 }
-            };
-            if let Some(help_msg) = help_msg {
-                diag.help(help_msg);
             }
+
+            decorate(&mut diag, span);
+
             diag.emit();
         },
     );
 }
 
-/// Reports a resolution failure diagnostic.
-///
-/// If we cannot find the exact source span of the resolution failure, we use the span of the
-/// documentation attributes themselves. This is a little heavy-handed, so we display the markdown
-/// line containing the failure as a note as well.
 fn resolution_failure(
     cx: &DocContext<'_>,
     item: &Item,
     path_str: &str,
     dox: &str,
     link_range: Option<Range<usize>>,
 ) {
-    build_diagnostic(
+    report_diagnostic(
         cx,
+        &format!("unresolved link to `{}`", path_str),
         item,
-        path_str,
         dox,
         link_range,
-        "cannot be resolved, ignoring it.",
-        "cannot be resolved, ignoring",
-        Some("to escape `[` and `]` characters, just add '\\' before them like `\\[` or `\\]`"),
+        |diag, sp| {
+            if let Some(sp) = sp {
+                diag.span_label(sp, "unresolved link");
+            }
+
+            diag.help(
+                r#"to escape `[` and `]` characters, just add '\' before them like `\[` or `\]`"#,
```

Comment:
While we're cleaning up things, I think 'just' is a little dismissive.
```suggestion
                r#"to escape `[` and `]` characters, add '\' before them: `\[` or `\]`"#,
```

Reply:
Fixed. What I'd really like to do here is use a structured suggestion but I don't think that's feasible until https://github.com/raphlinus/pulldown-cmark/issues/373 is fixed.